site/content/blog/2020-06-10-opentracing.md

```diff
@@ -0,0 +1,576 @@
+
+There are a few key terms used in tracing:
+
+1. **Trace** - a recording of the execution path of a request
+2. **Span** - a named, timed operation representing a contiguous segment inside the trace
+3. **Root Span** - the first span in a trace, a common ancestor to all spans in a trace
+4. **Context** - information identifying the request, required to connect spans in a distributed trace
+
+A trace recording usually looks something like this:
+
+![trace image](/blog/2020-06-10-opentracing/trace.png) 
+
+To add distributed tracing capabilities to [mattermost-server](https://github.com/mattermost/mattermost-server) we've picked [OpenTracing Go](https://github.com/opentracing/opentracing-go). We've already dove a little bit into Open Tracing implementation in the [first post of these series]({{< relref "./2019-10-25-instrumenting-go-code-via-ast.md" >}}).
+
+In this article we'll discuss all the nitty-gritty details of implementing a tracing system in your Go application without littering your code with repetitive, boilerplate tracing code.
+
+## The Goal
+
+So what are we actually working on? We want to make sure that every API request that is being handled by our server will get recorded into a trace, together with context information and ability to dive deep into the execution and allow easy problem analysis.
+
+The resulting system trace will look like this (using [Jaeger](https://www.jaegertracing.io/) web-ui visualization):
+![jaeger view](/blog/2020-06-10-opentracing/result.png) 
+
+## Naive approach
+
+To add tracing to any API call, we can do the following in our `ServeHTTP` function:
+
+```go
+func (h Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	c := &Context{}
+	// Start root span
+	span, ctx := tracing.StartRootSpanByContext(context.Background(), "apiHandler")
+	// Populate different span fields based on request headers
+	carrier := opentracing.HTTPHeadersCarrier(r.Header)
+	_ = opentracing.GlobalTracer().Inject(span.Context(), opentracing.HTTPHeaders, carrier)
+	ext.HTTPMethod.Set(span, r.Method)
+	ext.HTTPUrl.Set(span, c.App.Path())
+	ext.PeerAddress.Set(span, c.App.IpAddress())
+	span.SetTag("request_id", c.App.RequestId())
+	span.SetTag("user_agent", c.App.UserAgent())
+	// On handler exit: a) in case of an error, add it to the trace b) finish the span
+	defer func() {
+		if c.Err != nil {
+			span.LogFields(spanlog.Error(c.Err))
+			ext.HTTPStatusCode.Set(span, uint16(c.Err.StatusCode))
+			ext.Error.Set(span, true)
+		}
+		span.Finish()
+	}()
+	// Set current context to the one we got from root span, it will be passed down to actual api handlers
+	c.App.SetContext(ctx)
+	// ...
+	// Execute the actual API handler
+	h.HandleFunc(c, w, r)
+}
+```
+
+Next, we'll modify the actual API handler to nest it inside the parent span (we'll use `SearchUsers` as example)
+
+```go
+func (a *App) SearchUsers(props *model.UserSearch, options *model.UserSearchOptions) ([]*model.User, *model.AppError) {
```

Comment:
I was slightly confused by the way this is worded, `SearchUsers` intended to be a handler. But it is actually an app layer method, not in api layer. It should have the signature of `c *Context, w http.ResponseWriter, r *http.Request` if it's in the api layer. We should clarify that.